File: C++/005.cpp
/**
 * <p>$2520$ is the smallest number that can be divided by each of the numbers
 * from $1$ to $10$ without any remainder.</p> <p>What is the smallest positive
 * number that is <strong class="tooltip">evenly divisible<span
 * class="tooltiptext">divisible with no remainder</span></strong> by all of the
 * numbers from $1$ to $20$?</p>
 */

#include "../../../../usr/include/limits.h"

#include <iostream>

using ull = unsigned long long;

template <class T> bool divisibleByAll(T n) {
  for (int i = 20; i > 0; i--) {
    if (n % i != 0) {
      return false;
    }
  }
  return true;
}

int main() {
  ull i = 20 * 19;
  while (!divisibleByAll(i)) {
    i++;
  }
  std::cout << i;
  return 0;
}

File: typescript/005.ts
/**
 * <p>$2520$ is the smallest number that can be divided by each of the numbers
 * from $1$ to $10$ without any remainder.</p> <p>What is the smallest positive
 * number that is <strong class="tooltip">evenly divisible<span
 * class="tooltiptext">divisible with no remainder</span></strong> by all of the
 * numbers from $1$ to $20$?</p>
 */

function divisible_by_all_to_ceiling(ceiling: number): number {
  function bool_helper(n: number): boolean {
    for (let i: number = ceiling; i > 0; --i) {
      if (n % i !== 0) {
        return false;
      }
    }
    return true;
  }

  let i: number = ceiling * (ceiling - 1);
  while (!bool_helper(i)) {
    ++i;
  }
  return i;
}

console.log(divisible_by_all_to_ceiling(20));
